README.md
# Raytracing

<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-3-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

![julia-version](https://img.shields.io/badge/julia_version-v1.6-9558B2?style=flat&logo=julia) 
![package-version](https://img.shields.io/github/v/release/foglienimatteo/Raytracing?include_prereleases)
![CI-build](https://img.shields.io/github/workflow/status/foglienimatteo/Raytracing/Unit%20tests)
![size](https://img.shields.io/github/repo-size/foglienimatteo/Raytracing) 
![license]( https://img.shields.io/github/license/foglienimatteo/Raytracing)
[![codecov](https://codecov.io/gh/foglienimatteo/Raytracing/branch/master/graph/badge.svg?token=67GIZ9RA8Y)](https://codecov.io/gh/foglienimatteo/Raytracing)
[![Stable](https://img.shields.io/badge/docs-stable-blue.svg)](https://foglienimatteo.github.io/Raytracing/stable)
[![Dev](https://img.shields.io/badge/docs-dev-blue?style=flat)](https://foglienimatteo.github.io/Raytracing/dev)  

This software is a simple raytracing program written in the [Julia Programming Language](https://julialang.org).
It's based on the lectures of the [*Numerical techniques for photorealistic image generation*]( https://www.unimi.it/en/education/degree-programme-courses/2021/numerical-tecniques-photorealistic-image-generation) course (AY2020-2021), held by Associate Professor [Maurizio Tomasi](http://cosmo.fisica.unimi.it/persone/maurizio-tomasi) at University of Milan [Department of
Physics "Aldo Pontremoli"](http://eng.fisica.unimi.it/ecm/home).

</br>

## Table of Contents

- [Raytracing](#raytracing)
  - [Table of Contents](#table-of-contents)
  - [Installation](#installation)
  - [Demo and Demo Animation](#demo-and-demo-animation)
  - [Usage from the Command Line Interface](#usage-from-the-command-line-interface)
  - [Usage from the REPL](#usage-from-the-repl)
  - [Licence](#licence)
  - [Contributors ✨](#contributors-)

</br>

## Installation

The simplest way to install this software is cloning the repository where it is built in. Run in the command line

```bash
git clone https://github.com/foglienimatteo/Raytracing
```

or download the source code from the github repository https://github.com/foglienimatteo/Raytracing.

</br>

## Demo and Demo Animation

To start off and check the correct behavior of this software run one of the following command inside the main directory

```bash
./Raytracer.jl demo_animation --camera_type=per --width=640 --height=480 flat
```

```bash
./Raytracer.jl demo --world_type=B --camera_type=per --camera_position="[-1, 0, 1]" --width=640 --height=480 --samples_per_pixel=9 pathtracer 
```

and enjoy respectively the animation `demo/demo_animation_FHD.mp4` and the image `demo/demo_B_PathTracing_FHD.png` (these are renders of better quality then those specified in the two CLI commands above, you can compare your results with [demo_anim_Flat_640x480x360.mp4](demo/demo_anim_Flat_640x480x360.mp4) and [demo_B_PathTracing_640x480.png](demo/demo_B_PathTracing_640x480.png))

<!---
<video width="640" height="480"  type="video/mp4" "src="https://user-images.githubusercontent.com/79974922/119556147-ef2b3200-bd9e-11eb-956f-17de6ea6bdda.mp4"  autoplay loop> </video>"
-->

Demo Animation A  with FlatRenderer       | Demo Image B with PathTracing
:----------------------------------------:|:-------------------------:
![](demo/demo_animation_FHD.gif)  |  ![](demo/demo_B_PathTracing_FHD.png)

It may takes few minutes to render the animation; you might also give smaller (integer and even) values to `--width` and `--height` in order to obtain the same animation in a smaller amount of time (the price to pay is a worse definition of the animation itself).

</br>

## Usage from the Command Line Interface

This software is able to read a file that describes a scene (i.e a set of objects, pigments, materials, etc. that we want render).
To understand how to write such a file, take a look at the files contained in the [examples](esamples) directory, in the following order:

- [tutorial_basic_syntax.txt](examples/tutorial_basic_syntax.txt), which shows the basic syntax of how the scenefile must be written;
- [demo_world_B.txt](examples/demo_world_B.txt), which shows the rendering of the world type B of the `demo`function;
- [earth_and_sun.txt](examples/earth_and_sun.txt), which shows the rendering of a more complicated image with the point-light renderer;
- [earth_moon_sun.txt](examples/earth_moon_sun.txt), which shows the rendering of an animation.
  
Each of these files contains also, in the initial description, the commands necessary to render the described image; if you want to learn this software in the most practical way, these files are perfect!

If you are looking for a more complete and general documentation about how to write your text file and the CLI/repl options, you can find it in [DOCUMENTATION.md](./DOCUMENTATION.md).

The basic structure of a command in the CLI is the following:

```bash
./Raytracer.jl render [OPTIONS_FOR_THE_IMAGE] NAME_OF_THE_SCENEFILE {onoff|flat|pathtracer|pointlight}[OPTIONS_FOR_THE_RENDERER]
```

```bash
./Raytracer.jl animation --function=FUNCTION_NAME  --vec_variables="[NAMEVAR1, NAMEVAR2, ...]" --iterable=ITERABLE [OPTIONS_FOR_THE_IMAGE] NAME_OF_THE_SCENEFILE {onoff|flat|pathtracer|pointlight}[OPTIONS_FOR_THE_RENDERER]
```

There are four possible rendering algorithms; each of them is linked to different rules for the color of a pixel and the light ray that starts from that pixel and hits (or not) an object of the rendered scene:

- `onoff` : each pixel is set to the `background_color` if no shape is hit by its light ray, otherwise is set to `color`; this renderer exists only for debugging purposes.

- `flat` : each pixel is set to the `background_color` if no shape is hitten by its light ray, otherwise is set to the color of the hitted shape point.
  This renderer is very efficient, but it does not solve the rendering equation, and consequently no shadows or brightness are rendered.

- `pathtracer` : this is the TRUE renderer. It solves the rendering equation in the "standard" way, and for this reason it's very demanding.
  Nevertheless, the rendered images are incomparably more realistic than the ones made with the other three renderers; USE THIS RENDERER WITH FORESIGHT!

- `pointlight` : it's the "cheap solution" for a realistic image. This renderer creates an image setting each pixel colored or not depending on the "line of sight" between that point and the point-light sources in the scene.
  It's very efficient, and the images rendered are perfect for an astrophysical context or for very bright days of summer. Nevertheless it's a simple solution in order to avoid the longer times needed for the pathtracer algorithm.

The resulting files created at the end of the rendering are three:

- the PFM image (`scene.pfm` is the default name, if none is specified from the command line)
- the LDR image (`scene.png` is the default name, if none is specified from the command line)
- the JSON file, that saves some datas about input commands, rendering time etc. It has the same name of the LDR image and `.json` extention, so `scene.json` is the default name.

The resulting files created at the end of the animation are intead two:

- the animation (`scene_animation.mp4` is the default name, if none is specified from the command line)
- the JSON file; it has the same name of the animation and `.json` extention, so `scene_animation.json` is the default name.
  
Probably, the LDR image will not be "correctly" converted with the standard values used in the `render` function to tone-map the PFM file; it's consequently appropriate
to manually apply the tone mapping algorithm to the PFM image!
The tonemapping command is simple:

```bash
./Raytracer.jl tonemapping [--normalization ALPHA] [-gamma GAMMA] [--avg_lum LUM]FILE_PFM_TO_BE_TONEMAPPED NAME_OF_THE_RESULTING_LDR
```

where `ALPHA` is the scaling factor for the normalisation process (default `a=0.18`), `GAMMA` is YOUR monitor gamma value (default `g=1.0`) and `LUM` is the average luminosity of the image (the default value is manually calculated for the image itself, but it's useful to specify manually this value for particularly dark images).
Choose your `ALPHA` and `GAMMA` values without any fear to try again! This algorithm is indeed by far more efficient and computationally cheaper than the rendering.

NOTE: you can also specify `ALPHA`, `GAMMA` and `LUM` in the render and animation command; it's fundamental for the latter case, because no "PFM-animation" that can be tonemapped exists!

Here we show an example of usage, which renders the [earth_and_sun.txt](examples/earth_and_sun.txt) file

```bash
./Raytracer.jl render examples/earth_and_sun.txt --width=1200 --height=900 pointlight --dark_parameter=0.03
```

and of the tone-map the resulting PFM (named "scene.pfm")

```bash
./Raytracer.jl tonemapping scene.pfm scene.png --normalization=0.18 --gamma=1.0 --avg_lum=0.03
```

earth_and_sun.txt with Point-Light Renderer | 
:------------------------------------------:|
![](examples/earth_and_sun.png) 

We show also an animation example of usage, which renders the [earth_moon_sun.txt](examples/earth_moon_sun.txt) file

```bash
./Raytracer.jl animation  --normalization=0.18 --gamma=1.0 --avg_lum=0.065 --width=1200 --height=900 --function=earth_moon_sun --vec_variables="[moon_x,  moon_y, moon_z,  moon_rotang, earth_rotang]" --iterable=1:200 examples/earth_moon_sun.txt pointlight --dark_parameter=0.25
```

earth_moon_sun.txt with PointLight | 
:---------------------------------:|
![](examples/earth_moon_sun.gif) 

Refer to the latest [stable documentation](https://foglienimatteo.github.io/Raytracing/stable) for explanation of the functions used.

</br>

## Usage from the REPL

If you prefer to use the Julia REPL, first of all you need to include the software with the known command in the REPL:

```julia
include("Raytracer.jl")
```

You can obviously visualize the options for each function thanks to the help option (by typing `?`, than write the function name) and set the parameters in a dictionary-like sintax.
The only main difference is that in order to specify the renderer algorithm, the key is the not-so-intuitive string `%COMMAND%` (due to the [`ArgParse.jl`](https://github.com/carlobaldassi/ArgParse.jl) package used for the command line parsing operation).

For example, in order to render the demo image B previously showed:

```julia
demo("world_type"=>"B", "camera_position"=>"[-1.0, 0.0, 1.0]", "samples_per_pixel"=>9, "%COMMAND%"=>"pathtracer")
```

Instead, for rendering the `examples/earth_and_sun.txt`

```julia
render("scenefile"=>"examples/earth_and_sun.txt", "width"=>1200, "height"=>900, "%COMMAND%"=>"pointlight", "pointlight"=>Dict("dark_parameter"=>0.03))
```

and for tone-map it

```julia
tone_mapping("infile"=>"scene.pfm", "outfile"=>"scene.png", "normalization"=>0.18, "gamma"=>1.0, "avg_lum"=>0.04)
```


CHANGELOG.md
# HEAD

## VERSION 1.2.0

- fixed bug in camera settings (conflicts between CLI command-default and file parsing)
- added Torus as new shape

## VERSION 1.1.1

- fixed a bug in the point-light renderer algorithm;

- modified the scenefile examples;
  
## VERSION 1.1.0

- added the `render_animation` function and improved the CLI; now it's possible to read an animation from a file; see PR[#19](https://github.com/cosmofico97/Raytracing/pull/19)

## VERSION 1.0.0

- added the `render` function and implemented lexer and parser; now it's possible to read a scene from a file and render it; see PR[#18](https://github.com/cosmofico97/Raytracing/pull/18)

- Added the shape AABB, see PR[#22](https://github.com/cosmofico97/Raytracing/pull/22)
  
- Added the shape Cube, see PR[#21](https://github.com/cosmofico97/Raytracing/pull/21)

- Added the shape Triangle, see PR[#20](https://github.com/cosmofico97/Raytracing/pull/20)

## VERSION 0.5.0

- Added a third demo world and modified `sphere_to_point_uv`

## VERSION 0.4.0

- Added point-light tracing algorithm, see PR[#17](https://github.com/cosmofico97/Raytracing/pull/17)

## VERSION 0.3.0

- Added antialiasing algorithm, see PR[#16](https://github.com/cosmofico97/Raytracing/pull/16)

## VERSION 0.2.0

- Feature: it's possible to create animations of `demo` (required [ffmpeg](https://www.ffmpeg.org) software)
  
- Implemented `demo` function: two possible world versions can be rendered

- Impleented an ortho-normal basis (ONB) generator algorithm based on [Duff et al. 2017](https://graphics.pixar.com/library/OrthonormalB/paper.pdf) 
  
- Implemented `PCG` random number generator (see Melissa E. O’Neill (2014), ["PCG: A Family of Simple Fast Space-Efficient Statistically Good Algorithms for Random Number Generation"](https://www.pcg-random.org/paper.html)

- Implemented Renderers: `OnOffRenderer`, `FlatRRenderer`, `PathTracer`

- Implemented BRDFs: `BRDF`, `DiffuseBRDF`, `SpecularBRDF`
  
- Implemented Pigments: `Pigment`, `UniformPigment`, `CheckeredPigment`, `ImagePigment`

- Implemented Shapes: `Shape`, `Sphere`, `World`, `Plane`
  
- First bug fixed in the code (!); see PR[#7](https://github.com/cosmofico97/Raytracing/pull/7#issue-630790415))
  
- Implemented interface from command line
  
- Implemented a first main interface

## VERSION 0.1.0

- Added feature: now can convert from a .pfm image to .png and .tiff one 

- Completed function for reading and writing a .pfm image file format
  

src/Demo.jl
# -*- encoding: utf-8 -*-
#
# The MIT License (MIT)
#
# Copyright © 2021 Matteo Foglieni and Riccardo Gervasoni
#

using Printf
using ProgressMeter, ProgressBars


"""
	first_world() :: World

Render the first world (identified with the string "A").

This world consists in a set of 10 spheres of equal radius 0.1:
8 brown spheres are placed at the verteces of a cube of side 1.0, one green-purple 
checked sphere is in the center of the lower cube face and another multi-colored sphere is
in the center of the left cube face.

See also: [`World`](@ref), [`demo`](@ref), [`demo_animation`](@ref)
"""
function first_world()
	material1 = Material(DiffuseBRDF(UniformPigment(RGB(0.7, 0.3, 0.2))))

   	material2 = Material(DiffuseBRDF(CheckeredPigment(RGB(0.2, 0.7, 0.3), 
	    											  RGB(0.3, 0.2, 0.7), 
									                  4) )	)


	sphere_texture = HDRimage(2, 2)
	set_pixel(sphere_texture, 0, 0, RGB(0.1, 0.2, 0.3))
    	set_pixel(sphere_texture, 0, 1, RGB(0.2, 0.1, 0.3))
	set_pixel(sphere_texture, 1, 0, RGB(0.3, 0.2, 0.1))
    	set_pixel(sphere_texture, 1, 1, RGB(0.3, 0.1, 0.2))

	material3 = Material(DiffuseBRDF(ImagePigment(sphere_texture)))

	# Create a world and populate it with a few shapes
	world = World()
	for x in [-0.5, 0.5], y in [-0.5, 0.5], z in [-0.5, 0.5]
		add_shape!(world,
				Sphere( 
					translation(Vec(x, y, z)) * scaling(Vec(0.1, 0.1, 0.1)),
					material1
				)
		)
	end

	# Place two other balls in the bottom/left part of the cube, so
	# that we can check if there are issues with the orientation of
	# the image
	add_shape!(
		world, 
		Sphere( 
			translation(Vec(0.0, 0.0, -0.5)) * scaling(Vec(0.1, 0.1, 0.1)),
			material2
		)
	)
	add_shape!(
		world, 
		Sphere( 
			translation(Vec(0.0, 0.5, 0.0)) * scaling(Vec(0.1, 0.1, 0.1)),
			material3
		)
	)

	add_light!(
		world, 
		PointLight(Point(-10.0, 10.0, 10.0), 
		RGB{Float32}(1.0, 1.0, 1.0))
	)

	return world
end


"""
	second_world() :: World

Render the second world (identified with the string "B").

This world consists in a checked x-y plane, a blue opaque 
sphere, a red reflecting sphere, and a green oblique reflecting plane, all
inside a giant emetting sphere.

See also: [`World`](@ref), [`demo`](@ref), [`demo_animation`](@ref)
"""
function second_world()
	world = World()

	sky_material = 
		Material(
			DiffuseBRDF(UniformPigment(RGB{Float32}(0., 0., 0.))),
			UniformPigment(RGB{Float32}(1.0, 0.9, 0.5)),
		)

	ground_material = 
		Material(
			DiffuseBRDF(
				CheckeredPigment(
					RGB{Float32}(0.3, 0.5, 0.1),
					RGB{Float32}(0.1, 0.2, 0.5),
				)
			)
		)

	sphere_material = 
		Material(DiffuseBRDF(UniformPigment(to_RGB(0, 128, 240))))
	triangle_material = 
		Material(DiffuseBRDF(UniformPigment(to_RGB(190, 24, 120))))
	mirror_material = 
		Material(SpecularBRDF(UniformPigment(to_RGB(232, 10, 10))))
	mirror_material_2 = 
		Material(SpecularBRDF(UniformPigment(to_RGB(178, 255, 102))))
	
	
	add_shape!(
		world,
		Sphere(
			scaling(Vec(50, 50, 50)) * translation(Vec(0, 0, 0)),
			sky_material,
		)
	)
	
	add_shape!(
		world, 
		Plane(
			Transformation(), 
			ground_material,
		)
	)
	add_shape!(
		world, 
		Plane(
			translation(Vec(0., -2., 0)) * rotation_z(π/6.) * rotation_x(π/2.), 
			mirror_material_2,
		)
	)

	s1, s2 = 0.6, 1.0
	add_shape!(
		world,
		Sphere(
			translation(Vec(0, 0, 0.3)) * scaling(Vec(s1, s1, s1)),
			sphere_material,
		)
	)

	add_shape!(
		world,
		Sphere(
			translation(Vec(0.4, 1.5, 0)) * scaling(Vec(s2, s2, s2)),
			mirror_material,
		)
	)

	add_shape!(
	 	world, 
		Triangle( 
			Point(2.0, 1.0, 3.0), Point(3.0, 3.0, 1.0), Point(2.5, 0.0, 1.0),
			triangle_material
		)
	)

	add_shape!(
		world, 
		Cube( 
			translation(Vec(-0.3, -1.5, 0.3)) * scaling(Vec(s2, s2, s2)),
			mirror_material
		)
	)

	add_shape!(
		world,
		Torus(translation(Vec(0.4, -0.5, 2.2)) * rotation_y(pi/6) * rotation_x(pi/3) * scaling(Vec(0.2, 0.3, 0.12)),
		Material(DiffuseBRDF(CheckeredPigment(SYM_COL["PURPLE"], SYM_COL["NAVY"], 24))),
		1.0,
		3.0
		)
	)


	add_light!(
		world, 
		PointLight(Point(-1.0, 4.0, 2.0), 
		RGB{Float32}(1.0, 1.0, 1.0))
	)

	return world
end


function third_world()
	world = World()

	donut_material = Material(DiffuseBRDF(CheckeredPigment(to_RGB(128, 0, 128),
														   to_RGB(0, 255, 0),
														   8)),
							 CheckeredPigment(to_RGB(128, 0, 128), to_RGB(0, 255, 0),
														   8)
							)
    sphere_material = Material(DiffuseBRDF(UniformPigment(to_RGB(0, 128, 240))))
	# Create a world and populate it with a few shapes
	
	# add_shape!(world,
	# 		Torus( 
	# 			translation(Vec(20.0, 0, 0)) * scaling(Vec(0.1, 0.1, 0.1)),
	# 			donut_material,
	# 			1.0, 3.0
	# 		)
	# )

	add_shape!(
		world,
		Torus(#translation(Vec(0.4, 1.5, 2.)) * rotation_y(pi/4) * rotation_x(pi/4),
		translation(Vec(0.4, 1.5, 2.)) * rotation_z(pi/2),
			Material(donut_material.brdf, donut_material.emitted_radiance),
			1.0,
			3.0
		)
	)

	# s1, s2 = 0.6, 1.0
	# add_shape!(
	# 	world,
	# 	Sphere(
	# 		translation(Vec(0, 0, 0.3)) * scaling(Vec(s1, s1, s1)),
	# 		sphere_material,
	# 	)
	# )

	add_light!(
		world, 
		PointLight(Point(-10.0, 10.0, 10.0), 
		RGB{Float32}(255,255,255))
	)

	return world
end


"""
	select_world(type_world::String) ::Function

Select which demo world is used
"""
function select_world(type_world::String)
	(type_world=="A") && (return first_world())
	(type_world=="B") && (return second_world())
	(type_world=="C") && (return third_world())

	throw(ArgumentError("The input type of world $type does not exists"))
end

##########################################################################################92

function print_progress(row::Int64, col::Int64, height::Int64, width::Int64)
	print("Rendered row $(height - row)/$(height) \t= ")
	@printf "%.2f" 100*((height - row)/height)
	print("%\n")
end

function demo(x::(Pair{T1,T2} where {T1,T2})...)
	demo( parse_demo_settings(  Dict( pair for pair in [x...]) )... )
end

function demo(
		renderer::Renderer = FlatRenderer(),
		camera_type::String = "per",
		camera_position::Union{Point, Vec} = Point(-1.,0.,0.), 
     	α::Float64 = 0., 
     	width::Int64 = 640, 
     	height::Int64 = 480,
		a::Float64 = 0.18, 
        γ::Float64 = 1.0,
		lum::Union{Number, Nothing} = nothing,  
     	pfm_output::String = "demo.pfm", 
        png_output::String = "demo.png",
		samples_per_pixel::Int64 = 0, 
		world_type::String = "A",
		bool_print::Bool = true,
		bool_savepfm::Bool = true,
		ONLY_FOR_TESTS::Bool = false,
    )

    (ONLY_FOR_TESTS==false) || (return nothing)  

	# println("\n", world_type, "\n")
	renderer.world = select_world(world_type)

	samples_per_side = string2rootint64(string(samples_per_pixel))
	
	observer_vec = typeof(camera_position) == Point ?
		camera_position - Point(0., 0., 0.) :
		camera_position
	
	camera_tr = rotation_z(deg2rad(α)) * translation(observer_vec)
	aspect_ratio = width / height

	if camera_type == "per"
		(bool_print==true) && (println("Using perspective camera"))
		camera = PerspectiveCamera(1., aspect_ratio, camera_tr)
	elseif camera_type == "ort"
		(bool_print==true) && (println("Using orthogonal camera"))
		camera = OrthogonalCamera(aspect_ratio, camera_tr)
	else
		throw(ArgumentError("Unknown camera: $camera_type"))
	end

	if typeof(renderer) == OnOffRenderer
		(bool_print==true) && (println("Using on/off renderer"))
	elseif typeof(renderer) == FlatRenderer
		(bool_print==true) && (println("Using flat renderer"))
	elseif typeof(renderer) == PathTracer
		(bool_print==true) && (println("Using path tracing renderer"))
	elseif typeof(renderer) == PointLightRenderer
        (bool_print==true) && (println("Using point-light renderer"))
	else
		throw(ArgumentError("Unknown renderer: $(typeof(renderer))"))
	end

	# Run the ray-tracer
	image = HDRimage(width, height)
	tracer = ImageTracer(image, camera, samples_per_side)

	fire_all_rays!(tracer, renderer, (r,c) -> print_progress(r,c,image.height, image.width) )
	img = tracer.img

	# Save the HDR image
	(bool_savepfm==true) && (open(pfm_output, "w") do outf; write(outf, img); end)
	(bool_print==true) && (println("\nHDR demo image written to $(pfm_output)\n"))

	# Apply tone-mapping to the image
	if world_type=="A" && (isnothing(lum) || lum≈0.0)
		normalize_image!(img, a, 0.1)
		clamp_image!(img)
		γ_correction!(img, γ)
	else
		normalize_image!(img, a, lum)
		clamp_image!(img)
		γ_correction!(img, γ)
	end

	# Save the LDR image
	if (typeof(query(png_output)) == File{DataFormat{:UNKNOWN}, String})
		(bool_print==true) && (
			println(
				"File{DataFormat{:UNKNOWN}, String} for $(png_output)\n"*
				"Written as a .png file.\n"
			)
		)
     	Images.save(File{format"PNG"}(png_output), get_matrix(img))
	else
		Images.save(png_output, get_matrix(img))
	end

	(bool_print==true) && (println("\nHDR demo image written to $(png_output)\n"))
	nothing
end


"""
	demo(
        renderer::Renderer = FlatRenderer(),
		camera_type::String = "per",
		camera_position::Union{Point, Vec} = Point(-1.,0.,0.), 
     	α::Float64 = 0., 
     	width::Int64 = 640, 
     	height::Int64 = 480, 
		a::Float64=0.18, 
        γ::Float64=1.0, 
		lum::Union{Number, Nothing} = nothing,
     	pfm_output::String = "demo.pfm", 
        png_output::String = "demo.png",
		samples_per_pixel::Int64 = 0, 
		world_type::String = "A",
		bool_print::Bool = true,
		bool_savepfm::Bool = true,
		ONLY_FOR_TESTS::Bool = false,
        )

	demo(x::(Pair{T1,T2} where {T1,T2})...) = 
		demo( parse_demo_settings(  Dict( pair for pair in [x...]) )... )

Creates a demo image with the specified options. 

There are two possible demo image "world" to be rendered, specified through the
input string `type`.

The `type=="A"` demo image world consist in a set of 10 spheres of equal radius 0.1:
8 brown spheres are placed at the verteces of a cube of side 1.0, one green-purple 
checked sphere is in the center of the lower cube face and another multi-colored sphere is
in the center of the left cube face.

The `type=="B"` demo image world consists in a checked x-y plane, a blue opaque 
sphere, a red reflecting sphere, and a green oblique reflecting plane, all
inside a giant emetting sphere.

The creation of the demo image has the objective to check the correct behaviour of
the rendering software, specifically the orientation upside-down and left-right.

## Arguments

The following input arguments refers to the first method presented in the signature;
it's obviously very uncomfortable to use that method, so it's recommended to take 
advantage of the second one, which allows to write the input values in a dictionary
like syntax with arbitrary order and comfort. See the documentation of  
[`parse_demo_settings`](@ref) to learn how to use the keys:

- `renderer::Renderer = FlatRenderer()` : renderer to be used in the rendering, with all
  the settings already setted (exception made for the `world`, that will be overridden
  and created here)

- `camera_type::String = "per"` : set the perspective projection view:
  - `camera_type=="per"` -> set [`PerspectiveCamera`](@ref)  (default value)
  - `camera_type=="ort"`  -> set [`OrthogonalCamera`](@ref)

- `camera_position::Union{Point, Vec} = Point(-1.,0.,0.)` : set the point of observation 
  in (`X`,`Y,`Z`) coordinates, as a `Point`or a `Vec` input object.

- `α::Float64 = 0.` : angle of rotation _*IN RADIANTS*_, relative to the vertical
  (i.e. z) axis with a right-handed rule convention (clockwise rotation for entering (x,y,z)-axis 
  corresponds to a positive input rotation angle)

- `width::Int64 = 640` and `height::Int64 = 480` : pixel dimensions of the demo image;
  they must be both even positive integers.

- `a::Float64 = 0.18` : normalization scale factor for the tone mapping.

- `γ::Float64 = 1.27` : gamma factor for the tone mapping.

- `lum::Union{Number, Nothing} = nothing ` : average luminosity of the image; iIf not specified or equal to 0, 
  it's calculated through [`avg_lum`](@ref)

- `pfm_output::String = "demo.pfm"` : name of the output pfm file

- `png_output::String = "demo.png"` : name of the output LDR file

- `samples_per_pixel::Int64 = 0` : number of rays per pixel to be used (antialiasing);
  it must be a perfect square positive integer (0, 1, 4, 9, ...) and if is set to
  0 (default value) is choosen, no anti-aliasing occurs, and only one pixel-centered 
  ray is fired for each pixel.

- `world_type::String = "A"` : specifies the type of world to be rendered ("A" or "B")

- `bool_print::Bool = true` : specifies if the WIP messages of the demo
  function should be printed or not (useful option for [`demo_animation`](@ref))

- `bool_savepfm::Bool = true` : bool that specifies if the pfm file should be saved
  or not (useful option for [`demo_animation`](@ref))

- `ONLY_FOR_TESTS::Bool = false` : it's a bool variable conceived only to
  test the correct behaviour of the renderer for the input arguments; if set to `true`, 
  no rendering is made!

See also: [`Point`](@ref), [`Vec`](@ref), [`Renderer`](@ref) [`OnOffRenderer`](@ref), 
[`FlatRenderer`](@ref), [`PathTracer`](@ref), [`PointLightRenderer`](@ref),
[`demo_animation`](@ref), [`parse_demo_settings`](@ref)
""" 
demo

##########################################################################################92


function demo_animation(x::(Pair{T1,T2} where {T1,T2})...)
	demo_animation( parse_demoanimation_settings(  Dict( pair for pair in [x...]) )... )
end

function demo_animation(
			renderer::Renderer = FlatRenderer(),
			camera_type::String = "per",
			camera_position::Union{Point, Vec} = Point(-1.,0.,0.), 
        	width::Int64 = 200, 
        	height::Int64 = 150,
			a::Float64=0.18, 
            γ::Float64=1.0,
			lum::Union{Number, Nothing} = nothing,
       		anim_output::String = "demo-animation.mp4",
			samples_per_pixel::Int64 = 0,
			world_type::String = "A", 
			ONLY_FOR_TESTS::Bool = false,
		)

	run(`rm -rf .wip_animation`)
	run(`mkdir .wip_animation`)
	
	dict_gen = Dict(
			"camera_type"=>camera_type,
			"camera_position"=>camera_position,
			"renderer"=>renderer, 
			"width"=>width,
			"height"=>height,
			"normalization"=>a,
			"gamma"=>γ,
			"avg_lum"=>lum,
			"samples_per_pixel"=>samples_per_pixel,
			"world_type" => world_type,
			"set_pfm_name"=>".wip_animation/demo.pfm",
			"bool_print"=>false,
			"bool_savepfm"=>false,
			"ONLY_FOR_TESTS"=>ONLY_FOR_TESTS,
			)

	(ONLY_FOR_TESTS==false) || (return nothing)

	iter = ProgressBar(0:359)
	for angle in iter
		angleNNN = @sprintf "%03d" angle
		dict_spec = Dict(
					"alpha"=>1.0*angle,
					"set_png_name"=>".wip_animation/image$(angleNNN).png"
					)
		demo(parse_demo_settings(merge(dict_gen, dict_spec))...)
		set_description(iter, string(@sprintf("Frame generated: ")))
	end

	# -r 25: Number of frames per second
	run(`ffmpeg -r 25 -f image2 -s $(width)x$(height) -i 
	.wip_animation/image%03d.png -vcodec libx264 
	-pix_fmt yuv420p $(anim_output)`)

	run(`rm -rf .wip_animation`)
end

"""
	demo_animation( 
			renderer::Renderer = FlatRenderer(),
			camera_type::String = "per",
        	width::Int64 = 200, 
        	height::Int64 = 150,
       		anim_output::String = "demo-animation.mp4",
			samples_per_pixel::Int64 = 0,
			world_type::String = "A", 
			ONLY_FOR_TESTS::Bool = false,
		)

	demo_animation(x::(Pair{T1,T2} where {T1,T2})...) = 
		demo_animation( parse_demoanimation_settings(  Dict( pair for pair in [x...]) )... )

	
Creates an animation of the demo image with the specified options. It's
necessary to have istalled the ffmpeg software to run this function.

This function works following this steps:
- creates an hidden directory, called ".wip_animation"; if it already exists,
  it will be destroyed and recreated.
- inside ".wpi_animation", creates 360 png images of the demo image (using the 
  [`demo`](@ref) function with the specified projection, renderer and image 
  dims); each image correspons to a frame of the future animation
- through the `ffmpeg` software, the 360 png images are converted into the
  animation mp4 file, and saved in the main directory
- the ".wpi_animation" directory and all the png images inside it are destroyed


## Arguments

The following input arguments refers to the first method presented in the signature;
it's obviously very uncomfortable to use that method, so it's recommended to take 
advantage of the second one, which allows to write the input values in a dictionary
like syntax with arbitrary order and comfort. See the documentation of  
[`parse_demoanimation_settings`](@ref) to learn how to use the keys:

- `renderer::Renderer = FlatRenderer()` : renderer to be used in the rendering, with all
  the settings already setted (exception made for the `world`, that will be overridden
  and created here)

- `camera_type::String = "per"` : set the perspective projection view:
  - `camera_type=="per"` -> set [`PerspectiveCamera`](@ref)  (default value)
  - `camera_type=="ort"`  -> set [`OrthogonalCamera`](@ref)

- `camera_position::Union{Point, Vec} = Point(-1.,0.,0.)` : set the point of observation 
  in (`X`,`Y,`Z`) coordinates, as a `Point`or a `Vec` input object.

- `width::Int64 = 640` and `height::Int64 = 480` : pixel dimensions of the demo image;
  they must be both even positive integers.

- `a::Float64 = 0.18` : normalization scale factor for the tone mapping.

- `γ::Float64 = 1.27` : gamma factor for the tone mapping.

- `lum::Union{Number, Nothing} = nothing ` : average luminosity of the image; iIf not specified or equal to 0, 
  it's calculated through [`avg_lum`](@ref)

- `anim_output::String = "demo-animation.mp4"` : name of the output animation file

- `samples_per_pixel::Int64 = 0` : number of rays per pixel to be used (antialiasing);
  it must be a perfect square positive integer (0, 1, 4, 9, ...) and if is set to
  0 (default value) is choosen, no anti-aliasing occurs, and only one pixel-centered 
  ray is fired for each pixel.

- `world_type::String = "A"` : specifies the type of world to be rendered ("A" or "B")

- `ONLY_FOR_TESTS::Bool = false` : it's a bool variable conceived only to
  test the correct behaviour of the renderer for the input arguments; if set to `true`, 
  no rendering is made!

See also: [`Point`](@ref), [`Vec`](@ref), [`Renderer`](@ref) [`OnOffRenderer`](@ref), 
[`FlatRenderer`](@ref), [`PathTracer`](@ref), [`PointLightRenderer`](@ref),
[`demo`](@ref), [`parse_demoanimation_settings`](@ref)
"""
demo_animation


#=
for angle in $(seq 0 359); do
    angleNNN=$(printf "%03d" $angle)
    ./main demo --per --width=640 --height=480 --alpha=$angle --set-png-name="animazione/image${angleNNN}.png"
done

# -r 25: Number of frames per second
ffmpeg -r 25 -f image2 -s 50x30 -i animazione/image%03d.png \
    -vcodec libx264 -pix_fmt yuv420p \
    spheres-perspective.mp4

ffmpeg -i demo/demo_anim_Flat_640x480x360.mp4 -t 14 
	-pix_fmt rgb24 demo/demo_anim_Flat_640x480x360.gif

ffmpeg -i file.mp4 -pix_fmt rgb24 file.gif
=#
